### PR TITLE
feat: perform more chat token checks

### LINKED ISSUE #605: Chat should log a warning on token with missing chat:read scope
### Prerequisites

- [x] If this is a question/suggestion, I have already considered discussing it on the [Discord Server](https://discord.gg/FQ5vgW3)
- [X] I have attempted to look for similar github issues already

### Expected Behavior

I expect to be able to connect to a chatroom and receive message events. I am using the following code.

```java
        final TwitchIdentityProvider tip = new TwitchIdentityProvider(clientId, clientSecret, redirectUrl);
        authHandler = new AuthHandler(tip);


        CredentialManager credentialManager = CredentialManagerBuilder.builder().build();
        credentialManager.registerIdentityProvider(tip);

        MessageEventHandler messageEventHandler = new MessageEventHandler();

        twitchClient = TwitchClientPoolBuilder.builder()
                .withClientId(clientId)
                .withClientSecret(clientSecret)
                .withRedirectUrl(redirectUrl)
                .withCredentialManager(credentialManager)
                .withChatAccount(authHandler.getBotCredentials())
                .withDefaultAuthToken(authHandler.getBotCredentials())
                .withEnableChat(true)
                .withEnableHelix(true)
                .withEnableChatPool(true) // I have tried removing this line
                .withScheduledThreadPoolExecutor(authHandler.getExec())
                .withDefaultEventHandler(ReactorEventHandler.class) // I have tried removing this line
                .build();

        twitchClient.getChat().joinChannel("bendixbot");
        System.out.println("In BendixBot: " + twitchClient.getChat().isChannelJoined("bendixbot")); // Prints true
        System.out.println("In non-existing channel: " + twitchClient.getChat().isChannelJoined("bendixbottttt")); // Prints false

        // This line is copy/paste from your guide over at https://twitch4j.github.io/docs/events/channel-message-event/
        twitchClient.getChat().getEventManager().onEvent(ChannelMessageEvent.class, event -> {
            System.out.println("[" + event.getChannel().getName() + "] " + event.getUser().getName() + ": " + event.getMessage());
        });
```

### Current Behavior

I expect the examples to work as they're written and get some chat output 😋 There's not too much to say here.

### Log output / Stacktrace

```txt
Starting BendixBot...
Loading underlying JDBC driver.
Done.
Setting up data source.
Loading class `com.mysql.jdbc.Driver'. This is deprecated. The new driver class is `com.mysql.cj.jdbc.Driver'. The driver is automatically registered via the SPI and manual loading of the driver class is generally unnecessary.
Data source has been set up.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
In BendixBot: true
In non-existing channel: false
Started BendixBot in 5536ms
```

### Environment

- **Library Version**: 1.10.0
- **IDE**: IntelliJ IDEA 2021.3.2
- **Java Version**: 1.8.0_144
- **Build Tools**: Maven 3.8.1
- **Operating System**: Windows 10 64-bit

- **events4j-handler-reactor**: 0.10.0

### Build Configuration

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>dk.rasmusbendix</groupId>
    <artifactId>BendixBot</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>8</maven.compiler.source>
        <maven.compiler.target>8</maven.compiler.target>
    </properties>

    <dependencies>

        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>1.18.24</version>
        </dependency>


        <dependency>
            <groupId>com.github.twitch4j</groupId>
            <artifactId>twitch4j</artifactId>
            <version>1.10.0</version>
        </dependency>

        <!-- https://mvnrepository.com/artifact/org.apache.commons/commons-dbcp2 -->
        <dependency>
            <groupId>org.apache.commons</groupId>
            <artifactId>commons-dbcp2</artifactId>
            <version>2.9.0</version>
        </dependency>


        <dependency>
            <groupId>mysql</groupId>
            <artifactId>mysql-connector-java</artifactId>
            <version>8.0.29</version>
        </dependency>

        <dependency>
            <groupId>com.github.philippheuer.events4j</groupId>
            <artifactId>events4j-handler-reactor</artifactId>
            <version>0.10.0</version>
        </dependency>

        <dependency>
            <groupId>joda-time</groupId>
            <artifactId>joda-time</artifactId>
            <version>2.10.14</version>
        </dependency>

    </dependencies>

</project>
```

### Possible Solution

I don't have an exact idea, but MAXOUXAX mentioned that he thought it has been broken for months, if that helps troubleshooting it.

### Additional Context/Screenshots

As mentioned in the comments of my code in "Expected Behavior", I have tried changing a few of the settings for the TwitchClient. I also checked to see if the bot was in the channel I wanted it to be, and as my comments indicate, it is.

MAXOUXAX in the Discord server had a similar issue. He shared his bots logs which showed nothing out of the ordinary.
Our conversation on the Discord starts here: https://discord.com/channels/143001431388061696/268348117919858688/997259175178555472

### PR DESCRIPTION
**Pull request by iProdigy**
### Prerequisites for Code Changes
* [x] This pull request follows the code style of the project
* [x] I have tested this feature

### Issues Fixed 
* Closes #605

### Changes Proposed
On chat instantiation, add checks for:
* App access instead of user access passed
* Token without read scope passed
* Token without edit scope passed

On chat (re)connect:
* Add `validateTokenOnConnect` setting - will automatically connect in read-only mode if chat credential has expired

### CODE REVIEW

File: chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java [hunk 1]
```diff
@@ -451,20 +496,37 @@ protected void onConnected() {
             boolean sendRealPass = sendCredentialToThirdPartyHost // check whether this security feature has been overridden
                 || baseUrl.equalsIgnoreCase(TWITCH_WEB_SOCKET_SERVER) // check whether the url is exactly the official one
                 || baseUrl.equalsIgnoreCase(TWITCH_WEB_SOCKET_SERVER.substring(0, TWITCH_WEB_SOCKET_SERVER.length() - 4)); // check whether the url matches without the port
-            sendTextToWebSocket(String.format("pass oauth:%s", sendRealPass ? chatCredential.getAccessToken() : CryptoUtils.generateNonce(30)), true);
-            userName = String.valueOf(chatCredential.getUserName()).toLowerCase();
+
+            String token;
+            if (sendRealPass) {
+                if (validateOnConnect && connection.getConfig().backoffStrategy().getFailures() > 1 && !identityProvider.getAdditionalCredentialInformation(chatCredential).isPresent()) {
```
Comment (iProdigy) [inline chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java:502]: we need to call validate in a manner that allows us to distinguish network failure from expired token; currently, if there's a network issue, once it alleviates, we may end up reconnecting in anonymous mode, when the token wasn't ever invalid

File: chat/src/main/java/com/github/twitch4j/chat/TwitchChatBuilder.java [hunk 2]
```diff
@@ -257,6 +257,17 @@ public class TwitchChatBuilder {
...
+    /**
+     * Whether the {@link #getChatAccount()} should be validated on reconnection failures.
+     * <p>
+     * If enabled and the token has expired, chat will connect in read-only mode instead.
+     * <p>
+     * If the network connection is too slow, you may want to disable this setting to avoid
+     * disconnects while waiting for the result of the validate endpoint.
+     */
+    @With
+    private boolean validateReconnectToken = true;
```
Comment (iProdigy) [inline chat/src/main/java/com/github/twitch4j/chat/TwitchChatBuilder.java:269]: any better naming ideas?

Comment (PhilippHeuer) [reply chat/src/main/java/com/github/twitch4j/chat/TwitchChatBuilder.java:269]: as we provide the token in/as `chatAccount` how about `verifyChatAccountOnReconnect`?

File: auth/src/main/java/com/github/twitch4j/auth/providers/TwitchIdentityProvider.java [hunk 3]
```diff
@@ -38,6 +38,41 @@ public TwitchIdentityProvider(String clientId, String clientSecret, String redir
...
+                .build();
+
+            // perform call
+            Response response = HTTP_CLIENT.newCall(request).execute();
+
+            // return token status
+            if (response.isSuccessful())
+                return Optional.of(true);
+
+            if (response.code() >= 400)
```
Comment (PhilippHeuer) [inline auth/src/main/java/com/github/twitch4j/auth/providers/TwitchIdentityProvider.java:67]: invalid tokens always return 4xx, i think we should exclude 5xx here as those could be other unrelated issues.
> `if (response.code() >= 400 && response.code() < 500)`

Comment (iProdigy) [reply auth/src/main/java/com/github/twitch4j/auth/providers/TwitchIdentityProvider.java]: good point; resolved in https://github.com/twitch4j/twitch4j/pull/607/commits/f59935b422d92a58df31492185a632b1eae16a17

File: auth/src/main/java/com/github/twitch4j/auth/TwitchAuth.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public TwitchAuth(CredentialManager credentialManager, String clientId, String c
 
     public static void registerIdentityProvider(CredentialManager credentialManager, String clientId, String clientSecret, String redirectUrl) {
         // register the twitch identityProvider
-        Optional<OAuth2IdentityProvider> ip = credentialManager.getOAuth2IdentityProviderByName("twitch");
+        Optional<TwitchIdentityProvider> ip = credentialManager.getIdentityProviderByName("twitch", TwitchIdentityProvider.class);
         if (!ip.isPresent()) {
```
Comment (iProdigy) [inline auth/src/main/java/com/github/twitch4j/auth/TwitchAuth.java:37]: improbable breaking change here: if credential manager already has a `"twitch"` identity provider registered that isn't a `TwitchIdentityProvider`, then an exception will be thrown

File: chat/src/main/java/com/github/twitch4j/chat/TwitchChatBuilder.java [hunk 5]
```diff
@@ -284,16 +296,17 @@ public TwitchChat build() {
...
         // Register rate limits across the user id contained within the chat token
         final String userId;
         if (chatAccount == null) {
             userId = null;
         } else {
             if (StringUtils.isEmpty(chatAccount.getUserId())) {
-                chatAccount = credentialManager.getOAuth2IdentityProviderByName("twitch")
-                    .orElse(new TwitchIdentityProvider(null, null, null))
-                    .getAdditionalCredentialInformation(chatAccount).orElse(chatAccount);
+                credentialManager.getIdentityProviderByName("twitch", TwitchIdentityProvider.class)
```
Comment (iProdigy) [inline chat/src/main/java/com/github/twitch4j/chat/TwitchChatBuilder.java:307]: if we go back to using `orElse` here, we don't *need* `TwitchAuth` to throw exception on dupe provider name w/ different class

File: chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java [hunk 6]
```diff
@@ -302,18 +317,36 @@ public TwitchChat(WebsocketConnection websocketConnection, EventManager eventMan
...
+                }
+
+                // Check scopes
+                Collection<String> scopes = enriched.getScopes();
+                if (scopes.isEmpty() || (!scopes.contains(TwitchScopes.CHAT_READ.toString())) && !scopes.contains(TwitchScopes.KRAKEN_CHAT_LOGIN.toString())) {
+                    log.warn("TwitchChat: AccessToken does not have required scope ({}) to connect to chat, joining anonymously instead!", TwitchScopes.CHAT_READ);
+                    this.chatCredential = null; // connect anonymously to at least be able to read messages
+                }
+                if (!scopes.contains(TwitchScopes.CHAT_EDIT.toString())) {
+                    log.debug("TwitchChat: AccessToken does not have the scope to write messages ({}).", TwitchScopes.CHAT_EDIT);
```
Comment (PhilippHeuer) [inline chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java:348]: Should we log this as `warn` and append a "if you don't intend to write messages it's suggested to join anonymously"?

File: auth/src/main/java/com/github/twitch4j/auth/TwitchAuth.java (hunk 4 above)
Comment (PhilippHeuer) [reply auth/src/main/java/com/github/twitch4j/auth/TwitchAuth.java:37]: This change is fine imo, i think it is very unlikely that anyone did this.

File: auth/src/main/java/com/github/twitch4j/auth/providers/TwitchIdentityProvider.java [hunk 7]
```diff
@@ -38,6 +38,41 @@ public TwitchIdentityProvider(String clientId, String clientSecret, String redir
...
+
+        try {
+            // build request
+            Request request = new Request.Builder()
+                .url("https://id.twitch.tv/oauth2/validate")
+                .header("Authorization", "OAuth " + credential.getAccessToken())
+                .build();
+
+            // perform call
+            Response response = HTTP_CLIENT.newCall(request).execute();
```
Comment (PhilippHeuer) [inline auth/src/main/java/com/github/twitch4j/auth/providers/TwitchIdentityProvider.java:61]: These requests will possibly not use the auto configured proxy as other identity provider requests. The same applies to other requests in this class.

I added a httpClient in the OAuth2IdentityProvider which we can use.

File: chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java (hunk 6 above)
Comment (iProdigy) [reply chat/src/main/java/com/github/twitch4j/chat/TwitchChat.java:348]: we can raise it to at least `info`; i know there's folks who do this on purpose as they like their bot name showing up in many chatter lists

**Review by PhilippHeuer — APPROVED** [review]

**Pull request merged**